### PR TITLE
New dl esm info version

### PR DESCRIPTION
**Pull request by hiker**
This adds supported for the latest version of dl_esm_info (i.e. giving all public symbol a GO_ prefix).

I've added some (most) of the files from dl_esm_info to tests/test_files/gocean1.0/infrastructure, and added simple makefiles for compilation (of some of the gocean1.0 test files - obviously not all will compile) - though I have not added support for actually compiling tests.

### CODE REVIEW

File: src/psyclone/tests/test_files/gocean1p0/infrastructure/argument_mod.f90 [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+   ! The following value is valid for any go_arg.
+   enumerator :: GO_READ
+   ! The following values are only valid for fields.
+   enumerator :: GO_WRITE, GO_READWRITE, GO_INC
+   ! The following values are only valid for globals.
+   enumerator :: GO_MIN, GO_MAX, GO_SUM
+end enum
+
+! Fake array to enable us to provide stencil meta-data in the args() array
+integer, public, parameter :: stencil(0:111,0:111,0:111) = 0
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/infrastructure/argument_mod.f90]: I'm going to be slightly cheeky here and ask that you change to follow Rupert's suggestion (https://github.com/stfc/dl_esm_inf/issues/9) - assuming that compilers other than gfortran are happy with it.

File: src/psyclone/tests/test_files/gocean1p0/kernel_wrong_gridpt_type.f90 [hunk 2]
```diff
@@ -45,7 +45,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_wrong_gridpt_type.f90]: This should be `CU(I,J) = ...`

File: src/psyclone/tests/test_files/gocean1p0/kernel_wrong_access.f90 [hunk 3]
```diff
@@ -45,7 +45,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_wrong_access.f90]: This should still be `CU(I,J) = ...`

File: src/psyclone/tests/test_files/gocean1p0/kernel_stencil.f90 [hunk 4]
```diff
@@ -68,7 +68,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_stencil.f90]: This should still be `CU(I,J) = ...`

File: src/psyclone/tests/test_files/gocean1p0/kernel_ne_offset_mod.f90 [hunk 5]
```diff
@@ -109,7 +109,7 @@ subroutine compute_cv_code(i, j, cv, p, v)
     real(wp), intent(out), dimension(:,:) :: cv
     real(wp), intent(in),  dimension(:,:) :: p, v
 
-    CV(I,J) = .5d0*(P(I,J+1)+P(I,J))*V(I,J)
+    GO_CV(I,J) = .5d0*(P(I,J+1)+P(I,J))*V(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_ne_offset_mod.f90]: This should still be `CV(I,J) =  ...`

File: src/psyclone/tests/test_files/gocean1p0/kernel_missing_stencil.f90 [hunk 6]
```diff
@@ -39,7 +39,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_missing_stencil.f90]: This should still be `CU(I,J) = ...`

File: src/psyclone/tests/test_files/gocean1p0/kernel_missing_offset.f90 [hunk 7]
```diff
@@ -100,7 +100,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_missing_offset.f90]: This should still be `CU(I,J) = ...` and in commented-code above.

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_stencil.f90 [hunk 8]
```diff
@@ -39,7 +39,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_invalid_stencil.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_missing_iterates_over.f90 [hunk 9]
```diff
@@ -102,7 +102,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_missing_iterates_over.f90]: This should not have the GO_ suffix and in commented code above.

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_offset.f90 [hunk 10]
```diff
@@ -100,7 +100,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_invalid_offset.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_offset.f90 [hunk 11]
```diff
@@ -54,21 +54,21 @@ subroutine invoke_compute_cu(cufld, pfld, ufld)
...
 
-    ! Quantity CU is mass flux in x direction.
+    ! Quantity GO_CU is mass flux in x direction.
 
     ! Original code looked like:
     !
     !    DO J=1,N
     !      DO I=1,M
-    !           CU(I+1,J) = .5*(P(I+1,J)+P(I,J))*U(I+1,J)
+    !           GO_CU(I+1,J) = .5*(P(I+1,J)+P(I,J))*U(I+1,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_invalid_offset.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_iterates_over.f90 [hunk 12]
```diff
@@ -105,7 +105,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_invalid_iterates_over.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_grid_property.f90 [hunk 13]
```diff
@@ -43,7 +43,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_invalid_grid_property.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_fortran.f90 [hunk 14]
```diff
@@ -38,7 +38,7 @@ subroutine compute_cu_code(i, j, cu, p, u)
     real(wp), intent(out), dimension(:,:) :: cu
     real(wp), intent(in),  dimension(:,:) :: p, u
 
-    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+    GO_CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_invalid_fortran.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_iterates_over.f90 [hunk 15]
```diff
@@ -59,21 +59,21 @@ subroutine invoke_compute_cu(cufld, pfld, ufld)
     !  o  x  x  x
     !  o  x  x  x   j=1
 
-    ! Quantity CU is mass flux in x direction.
+    ! Quantity GO_CU is mass flux in x direction.
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_invalid_iterates_over.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_iterates_over.f90 [hunk 16]
```diff
@@ -59,21 +59,21 @@ subroutine invoke_compute_cu(cufld, pfld, ufld)
...
 
     ! Original code looked like:
     !
     !    DO J=1,N
     !      DO I=1,M
-    !           CU(I+1,J) = .5*(P(I+1,J)+P(I,J))*U(I+1,J)
+    !           GO_CU(I+1,J) = .5*(P(I+1,J)+P(I,J))*U(I+1,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_invalid_iterates_over.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_offset.f90 [hunk 17]
```diff
@@ -54,21 +54,21 @@ subroutine invoke_compute_cu(cufld, pfld, ufld)
     !  o  x  x  x
     !  o  x  x  x   j=1
 
-    ! Quantity CU is mass flux in x direction.
+    ! Quantity GO_CU is mass flux in x direction.
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_invalid_offset.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_missing_iterates_over.f90 [hunk 18]
```diff
@@ -56,21 +56,21 @@ subroutine invoke_compute_cu(cufld, pfld, ufld)
     !  o  x  x  x
     !  o  x  x  x   j=1
 
-    ! Quantity CU is mass flux in x direction.
+    ! Quantity GO_CU is mass flux in x direction.
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_missing_iterates_over.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_missing_iterates_over.f90 [hunk 19]
```diff
@@ -56,21 +56,21 @@ subroutine invoke_compute_cu(cufld, pfld, ufld)
...
 
     ! Original code looked like:
     !
     !    DO J=1,N
     !      DO I=1,M
-    !           CU(I+1,J) = .5*(P(I+1,J)+P(I,J))*U(I+1,J)
+    !           GO_CU(I+1,J) = .5*(P(I+1,J)+P(I,J))*U(I+1,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_missing_iterates_over.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_missing_offset.f90 [hunk 20]
```diff
@@ -54,21 +54,21 @@ subroutine invoke_compute_cu(cufld, pfld, ufld)
     !  o  x  x  x
     !  o  x  x  x   j=1
 
-    ! Quantity CU is mass flux in x direction.
+    ! Quantity GO_CU is mass flux in x direction.
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_missing_offset.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_missing_offset.f90 [hunk 21]
```diff
@@ -54,21 +54,21 @@ subroutine invoke_compute_cu(cufld, pfld, ufld)
...
 
     ! Original code looked like:
     !
     !    DO J=1,N
     !      DO I=1,M
-    !           CU(I+1,J) = .5*(P(I+1,J)+P(I,J))*U(I+1,J)
+    !           GO_CU(I+1,J) = .5*(P(I+1,J)+P(I,J))*U(I+1,J)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_missing_offset.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_ne_offset_mod.f90 [hunk 22]
```diff
@@ -63,21 +63,21 @@ subroutine invoke_compute_cv(cvfld, pfld, vfld)
     !  x  x  x  o
     !  o  o  o  o   j=1
 
-    ! Quantity CV is mass flux in y direction.
+    ! Quantity GO_CV is mass flux in y direction.
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_ne_offset_mod.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/kernel_ne_offset_mod.f90 [hunk 23]
```diff
@@ -63,21 +63,21 @@ subroutine invoke_compute_cv(cvfld, pfld, vfld)
...
 
     ! Original code looked like:
     !
     !    DO J=1,N
     !      DO I=1,M
-    !           CV(I,J+1) = .5*(P(I,J+1)+P(I,J))*V(I,J+1)
+    !           GO_CV(I,J+1) = .5*(P(I,J+1)+P(I,J))*V(I,J+1)
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/kernel_ne_offset_mod.f90]: This should not have the GO_ suffix.

File: src/psyclone/tests/test_files/gocean1p0/infrastructure/argument_mod.f90 (hunk 1 above)
Comment (hiker) [reply src/psyclone/tests/test_files/gocean1p0/infrastructure/argument_mod.f90]: Done. This needs the following merge request in dl_esm_info to be done first: https://github.com/stfc/dl_esm_inf/pull/11 to define the new go_stencil type.

File: src/psyclone/tests/test_files/gocean1p0/infrastructure/argument_mod.f90 [hunk 24]
```diff
@@ -0,0 +1,87 @@
+!-----------------------------------------------------------------------------
+! (c) The copyright relating to this information/data is jointly owned by
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/test_files/gocean1p0/infrastructure/argument_mod.f90]: That's a separate issue for dl_esm_inf (now that we have this as a submodule), I won't fix this here.

**Comment by hiker** [comment]
OK, I think this is ready to go again -  by now this patch is pretty big :)
I am not sure about the coverage reduction indicated above. Is there an easy way to see which lines are not covered anymore? It took me forever to find the one line that was not covered anymore (I compared the output of a random previous CI run, manually compared the line numbers that are not covered to find the one line number that has changed). I am sure there is an easier way :)

File: src/psyclone/tests/test_files/gocean1p0/infrastructure/argument_mod.f90 (hunk 1 above)
Comment (arporter) [reply src/psyclone/tests/test_files/gocean1p0/infrastructure/argument_mod.f90]: Thanks - I've just reviewed it.

**Comment by arporter** [comment]
Almost there now. Just the one missed correction to sort out and stfc/dl_esm_inf#11 to get done.

For coverage checking just follow the "Continue to review full report at Codecov." link in the CodeCov comment. I've had a look and there's just one line that's not covered and it's not a line I care about (in ghproto) so we're good :-)

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_fortran.f90 (hunk 14 above)
Comment (hiker) [reply src/psyclone/tests/test_files/gocean1p0/kernel_invalid_fortran.f90]: My apologies, I didn't know. I used it to keep track of what I had done ... and that obviously wasn't exactly too successful 😏 It seems that I somehow missed that file in my makefile, and so it was never compiled and tested *sigh* OK, I fixed that now - thanks!

**Comment by hiker** [comment]
OK, done. Thanks!

File: src/psyclone/tests/test_files/gocean1p0/kernel_invalid_fortran.f90 (hunk 14 above)
Comment (arporter) [reply src/psyclone/tests/test_files/gocean1p0/kernel_invalid_fortran.f90]: That's OK - how were you to know if we didn't tell you? :-) Thanks for fixing.

**Comment by arporter** [comment]
GitHub reports that the branch can be merged automatically.
All tests pass/xfail. All new code is covered.
We do not yet support compilation of generated code for the GOcean API.
Documentation builds OK but is not up-to-date - it still talks about the GOcean library on puma whereas that has been replaced by the dl_esm_inf submodule. However, since the addition of the submodule was my idea, I'll create a separate ticket to fix this. In fact, the documentation of the GOcean library that is currently in the PSyclone UG should actually be in dl_esm_inf.

**Comment by hiker** [comment]
OK, I am confused - I pulled master, and I can't see this in yet ;) Did you perhaps forget to push?

Anyway - if this is not my mistake, could you wait with merging this into master? I have started to port my application to use the new dl_esm_inf, and I found several bugs still in dl_esm_inf and psyclone, e.g.:
```
"grid_x_min_index": "simulation_domain%xstart
```
while simulation_domain has been replaced with ```subdomain%internal``` in dl_esm_inf. It passes all tests, since the tests check for simulation_domain. Being able to compile the created code would show this problem.

**Comment by arporter** [comment]
Hi @hiker, I was only thinking last night whether this might have caused problems - I was being paranoid and waiting for Travis to report before I hit the final 'merge' button. So, I can hold off, no problem...

**Comment by arporter** [comment]
Latest changes are all fine and are all covered. Source code and test code no longer has "simulation_domain" anywhere :-)

**Pull request merged**